### DISCUSSION TITLE
system.date() and system.time()

### DISCUSSION THREAD

**Discussion by kroggen** [aergoio/aergo#271] [General]
I am reviewing some functions that are not public yet to make them public (on documentation) soon

This is a place to discuss the functions `system.date()` and `system.time()`

**Comment by kroggen** [comment]
The above 2 functions are located on `system_module.c`

Apart from those, similar functions are on Aergo's LuaJIT (the module is disabled, but the functions are there)

Maybe the `lib_os.c` can be removed from Aergo's LuaJIT, or its content could be made unaccessible using some `#ifdef`

**Comment by kroggen** [comment]
Another thing to test is if there is any format specifier for date/time that uses local settings

The result must always be deterministic, regardless of location

**Reply by kroggen** [reply]
Many common modifiers are locale-dependent. Here are some examples:

%a: Abbreviated weekday name.
%A: Full weekday name.
%b: Abbreviated month name.
%B: Full month name.
%c: Local date and time representation.
%p: Either "AM" or "PM" based on the given time value. However, not all locales recognize this result.
%x: Local date representation.
%X: Local time representation.

**Reply by kroggen** [reply]
Here is the full list:

The specifiers marked with an asterisk (*) are locale-dependent

specifier | Replaced by | Example
-- | -- | --
%a | Abbreviated weekday name * | Thu
%A | Full weekday name * | Thursday
%b | Abbreviated month name * | Aug
%B | Full month name * | August
%c | Date and time representation * | Thu Aug 23 14:55:02 2001
%C | Year divided by 100 and truncated to integer (00-99) | 20
%d | Day of the month, zero-padded (01-31) | 23
%D | Short MM/DD/YY date, equivalent to %m/%d/%y | 08/23/01
%e | Day of the month, space-padded ( 1-31) | 23
%F | Short YYYY-MM-DD date, equivalent to %Y-%m-%d | 2001-08-23
%g | Week-based year, last two digits (00-99) | 01
%G | Week-based year | 2001
%h | Abbreviated month name * (same as %b) | Aug
%H | Hour in 24h format (00-23) | 14
%I | Hour in 12h format (01-12) | 02
%j | Day of the year (001-366) | 235
%m | Month as a decimal number (01-12) | 08
%M | Minute (00-59) | 55
%n | New-line character ('\n') |  
%p | AM or PM designation | PM
%r | 12-hour clock time * | 02:55:02 pm
%R | 24-hour HH:MM time, equivalent to %H:%M | 14:55
%S | Second (00-61) | 02
%t | Horizontal-tab character ('\t') |  
%T | ISO 8601 time format (HH:MM:SS), equivalent to %H:%M:%S | 14:55:02
%u | ISO 8601 weekday as number with Monday as 1 (1-7) | 4
%U | Week number with the first Sunday as the first day of week one (00-53) | 33
%V | ISO 8601 week number (01-53) | 34
%w | Weekday as a decimal number with Sunday as 0 (0-6) | 4
%W | Week number with the first Monday as the first day of week one (00-53) | 34
%x | Date representation * | 08/23/01
%X | Time representation * | 14:55:02
%y | Year, last two digits (00-99) | 01
%Y | Year | 2001
%z | ISO 8601 offset from UTC in timezone (1 minute=1, 1 hour=100)If timezone cannot be determined, no characters | +100
%Z | Timezone name or abbreviation *If timezone cannot be determined, no characters | CDT
%% | A % sign | %

**Reply by kroggen** [reply]
Here are the specifiers that are deterministic (not  locale-dependent)

specifier | Replaced by | Example
-- | -- | --
%c | Date and time representation (modified) | 2001-08-23 14:55:02
%C | Year divided by 100 and truncated to integer (00-99) | 20
%d | Day of the month, zero-padded (01-31) | 23
%D | Short MM/DD/YY date, equivalent to %m/%d/%y | 08/23/01
%e | Day of the month, space-padded ( 1-31) | 23
%F | Short YYYY-MM-DD date, equivalent to %Y-%m-%d | 2001-08-23
%g | Week-based year, last two digits (00-99) | 01
%G | Week-based year | 2001
%H | Hour in 24h format (00-23) | 14
%I | Hour in 12h format (01-12) | 02
%j | Day of the year (001-366) | 235
%m | Month as a decimal number (01-12) | 08
%M | Minute (00-59) | 55
%n | New-line character ('\n') |  
%p | AM or PM designation | PM
%R | 24-hour HH:MM time, equivalent to %H:%M | 14:55
%S | Second (00-61) | 02
%t | Horizontal-tab character ('\t') |  
%T | ISO 8601 time format (HH:MM:SS), equivalent to %H:%M:%S | 14:55:02
%u | ISO 8601 weekday as number with Monday as 1 (1-7) | 4
%U | Week number with the first Sunday as the first day of week one (00-53) | 33
%V | ISO 8601 week number (01-53) | 34
%w | Weekday as a decimal number with Sunday as 0 (0-6) | 4
%W | Week number with the first Monday as the first day of week one (00-53) | 34
%y | Year, last two digits (00-99) | 01
%Y | Year | 2001
%% | A % sign | %

**Reply by kroggen** [reply]
According to [this source](https://man7.org/linux/man-pages/man3/strftime.3.html) the `%p` modifier (AM/PM) depends on locale also. So it should be removed.

With that, it makes sense to not support `%I` as well (12-hour format)

**Reply by kroggen** [reply]
We can also remove `%t` and `%n` as these can be replaced with `\t` and `\n` on Lua

**Comment by kroggen** [comment]
The functions `system.date()` and `system.time()` have a field `isdst` used to mark Daylight Saving Time

But the time used on blockchain is always UTC. There is no local time supported, because all the nodes must return the same value regardless of their location.

So it appears reasonable to remove the `isdst` field from these functions

They are [here](https://github.com/aergoio/aergo/blob/develop/contract/system_module.c#L325) and [here](https://github.com/aergoio/aergo/blob/develop/contract/system_module.c#L366)

Is there any remark regarding this?

**Comment by hayarobi** [comment]
I think nearly all format specifiers can be enabled if we stick environments; locale to en (en_US?) and timezone to UTC, though I am only using ISO8601 format.

**Reply by kroggen** [reply]
The question is if they are really needed on chain. I suspect that the conversion from an UTC time can be made on client side, so it can use any locale and timezone. The chain can limit its support for UTC only.

The above is for cases when we read from blockchain.

There is one case that it would be good to have local support: when sending transactions and signing on hardware wallet, for example. If the user see one date on the dApp/browser and then it would see another date/time (UTC) when signing.

For this case it could support adding a timezone to the datetime, like `2023-11-13 10:24:30 -03:00`

**Reply by kroggen** [reply]
But the above can be done in the smart contract, by parsing the date, getting the time zone, and applying the difference to the timestamp
